src/http/http2.h

```diff
@@ -0,0 +1,604 @@
+        frame->hd.stream_id,
+        nghttp2_strerror(rc)));
+    }
+
+    return 0;
+  }
+
+  static int on_begin_headers_callback_client(
+    nghttp2_session* session, const nghttp2_frame* frame, void* user_data)
+  {
+    LOG_TRACE_FMT(
+      "http2::on_begin_headers_callback_client: {}", frame->hd.type);
+
+    return 0;
+  }
+
+  static int on_header_callback(
+    nghttp2_session* session,
+    const nghttp2_frame* frame,
+    const uint8_t* name,
+    size_t namelen,
+    const uint8_t* value,
+    size_t valuelen,
+    uint8_t flags,
+    void* user_data)
+  {
+    auto k = std::string(name, name + namelen);
+    auto v = std::string(value, value + valuelen);
+    LOG_TRACE_FMT("http2::on_header_callback: {}:{}", k, v);
+
+    auto* s = get_session(user_data);
+    auto* stream_data = get_stream_data(session, frame->hd.stream_id);
+
+    if (k == http2::headers::PATH)
+    {
+      stream_data->url = v;
+    }
+    else if (k == http2::headers::METHOD)
+    {
+      stream_data->verb = v;
+    }
+    else
+    {
+      stream_data->headers.emplace(k, v);
+    }
+
+    return 0;
+  }
+
+  static int on_header_callback_client(
+    nghttp2_session* session,
+    const nghttp2_frame* frame,
+    const uint8_t* name,
+    size_t namelen,
+    const uint8_t* value,
+    size_t valuelen,
+    uint8_t flags,
+    void* user_data)
+  {
+    auto k = std::string(name, name + namelen);
+    auto v = std::string(value, value + valuelen);
+    LOG_TRACE_FMT("http2::on_header_callback_client: {}:{}", k, v);
+
+    auto* stream_data = get_stream_data(session, frame->hd.stream_id);
+
+    if (k == http2::headers::STATUS)
+    {
+      // Note: handle errors
+      stream_data->status = HTTP_STATUS_OK;
+    }
+    else
+    {
+      stream_data->headers.emplace(k, v);
+    }
+
+    return 0;
+  }
+
+  static int on_data_callback(
+    nghttp2_session* session,
+    uint8_t flags,
+    StreamId stream_id,
+    const uint8_t* data,
+    size_t len,
+    void* user_data)
+  {
+    LOG_TRACE_FMT("http2::on_data_callback: {}", stream_id);
+
+    auto* stream_data = get_stream_data(session, stream_id);
+
+    stream_data->request_body.insert(
+      stream_data->request_body.end(), data, data + len);
+
+    return 0;
+  }
+
+  static int on_data_callback_client(
+    nghttp2_session* session,
+    uint8_t flags,
+    StreamId stream_id,
+    const uint8_t* data,
+    size_t len,
+    void* user_data)
+  {
+    LOG_TRACE_FMT("http2::on_data_callback_client: {}", stream_id);
+
+    auto* stream_data = get_stream_data(session, stream_id);
+
+    stream_data->response_body.insert(
+      stream_data->response_body.end(), data, data + len);
+
+    return 0;
+  }
+
+  static int on_stream_close_callback(
+    nghttp2_session* session,
+    StreamId stream_id,
+    uint32_t error_code,
+    void* user_data)
+  {
+    LOG_TRACE_FMT(
+      "http2::on_stream_close_callback: {}, {}", stream_id, error_code);
+
+    // Note: Stream should be closed correctly here
```

Comment:
Yes please! This might be related to the failures I saw with `dig`.

Reply:
I thought there was a specific message to send to the client on stream close but I believe there's isn't. This comment is about doing some cleanup on the server side but as we currently only support a single stream per session, the session will automatically be closed at this point so there's nothing to do. In other words, this shouldn't cause any issues with `dig`.